libp2p/muxers/mplex/coder.nim

```diff
@@ -0,0 +1,78 @@
+## Nim-LibP2P
+## Copyright (c) 2018 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import chronos, options, sequtils, strformat
+import nimcrypto/utils, chronicles
+import types,
+       ../../connection,
+       ../../varint,
+       ../../vbuffer,
+       ../../stream/lpstream
+
+logScope:
+  topic = "mplex-coder"
+
+type
+  Msg* = tuple 
+    id: uint 
+    msgType: MessageType
+    data: seq[byte]
+
+proc readMplexVarint(conn: Connection): Future[Option[uint]] {.async, gcsafe.} =
+  var
+    varint: uint
+    length: int
+    res: VarintStatus
+  var buffer = newSeq[byte](10)
+  result = none(uint)
+  try:
+    for i in 0..<len(buffer):
+      await conn.readExactly(addr buffer[i], 1)
+      res = LP.getUVarint(buffer.toOpenArray(0, i), length, varint)
+      if res == VarintStatus.Success:
+        return some(varint)
+    if res != VarintStatus.Success:
+      buffer.setLen(0)
+      return
+  except LPStreamIncompleteError:
```

Comment:
what if some other random exception happens in some of the calls? you have no idea what might come out of there, now or in a future refactoring

Reply:
Yeah, basically with exceptions you only handle what you know of and allow the exception to bubble up until other layers handle it or the process exits.

The idea being that when an exceptional condition happened it is either recoverable or non recoverable. Some exceptions are logical errors and might be recovered from, others aren't, that includes those exceptional conditions that aren't explicitly handled/known. 

This mechanism should prevent an application from ending up in an inconsistent state and exiting the process is prefered to running in this inconsistent state. 

That said, I agree that in general this model isn't perfect, just pointing out that this is by design and not an oversight.

Reply:
yes, but in this particular case, `conn.readExactly` will raise any number of exception types afair. won't the state be inconsistent in that case?

Reply:
ah nevermind, buffer is a local.

Reply:
which begs the question, why is it being reset? it doesn't get used after the exception - confused..

Reply:
good point, it's pointless resetting it there. I'll fix.